### PR TITLE
Add cursor support

### PR DESCRIPTION
**Pull request by fowles**
Adds cursor support and java support with conditional version checking for both.  I have not regenerated the documentation, because my experiments with vimdoc failed.

### CODE REVIEW

**Comment by fowles** [comment]
I signed it.

**Comment by malcolmr** [comment]
Generally looks reasonable, though:

- tests are failing :-)
- I think adding support for Java formatting with clang-format should be a separate discussion. (For example, I'm not sure how much it'd interfere with #43, or with anyone using a local formatter for java files.)

File: autoload/codefmt.vim [hunk 1]
```diff
@@ -186,19 +223,39 @@ function! codefmt#GetClangFormatFormatter() abort
...
       call maktaba#ensure#IsNumber(l:endline)
       let l:cmd += ['-lines', l:startline . ':' . l:endline]
     endfor
 
+    " Version 3.4 introduced support for cursor tracking
+    " http://llvm.org/releases/3.4/tools/clang/docs/ClangFormat.html
+    let l:supports_cursor = s:ClangFormatHasAtLeastVersion([3, 4])
+    if l:supports_cursor
+      " line2byte counts bytes from 1, and col counts from 1, so -2 
+      let l:cursor_pos = line2byte(line(".")) + col(".") - 2
```
Comment (dbarnett) [inline autoload/codefmt.vim]: Nit: single quotes.

File: autoload/codefmt.vim [hunk 2]
```diff
@@ -186,19 +223,39 @@ function! codefmt#GetClangFormatFormatter() abort
...
-    call maktaba#buffer#Overwrite(1, line('$'), l:formatted)
+
+    if !l:supports_cursor
+      call maktaba#buffer#Overwrite(1, line('$'), l:formatted[0:])
+    else
+      call maktaba#buffer#Overwrite(1, line('$'), l:formatted[1:])
+      try
+        let l:clang_format_output_json = maktaba#json#Parse(l:formatted[0])
+        let l:new_cursor_pos =
+              \ maktaba#ensure#IsNumber(l:clang_format_output_json.Cursor) + 1
```
Comment (dbarnett) [inline autoload/codefmt.vim]: Any scenarios where either no JSON gets returned or you get JSON w/o Cursor that you might want custom messages for?

Comment (fowles) [inline autoload/codefmt.vim]: my understanding is that the JSON is guaranteed to be there and that it is guaranteed to have cursor.  I just put this there to be extra double safe.

File: autoload/codefmt.vim [hunk 3]
```diff
@@ -186,19 +223,39 @@ function! codefmt#GetClangFormatFormatter() abort
...
+        exec "goto " . l:new_cursor_pos
```
Comment (dbarnett) [inline autoload/codefmt.vim]: Suggest `execute 'goto' l:new_cursor_pos`. (Long command name, no explicit concatenation.)

File: autoload/codefmt.vim (hunk 1 above)
Comment (fowles) [inline autoload/codefmt.vim]: Done.

File: autoload/codefmt.vim (hunk 3 above)
Comment (fowles) [inline autoload/codefmt.vim]: done

**Comment by fowles** [comment]
Not sure what you mean by "interfere with #43".  I think this implements it.  Custom formatters will take precedence, since we install the built-ins as "defaults" that come last in the lookup order.

**Comment by fowles** [comment]
fixed tests.  PTAL

**Comment by fowles** [comment]
pulled out the java stuff so we can handle it separately

File: autoload/codefmt.vim [hunk 4]
```diff
@@ -163,7 +191,7 @@ function! codefmt#GetClangFormatFormatter() abort
   function l:formatter.AppliesToBuffer() abort
     return &filetype is# 'c' || &filetype is# 'cpp' ||
          \ &filetype is# 'proto' || &filetype is# 'javascript'
-  endfunction
+  endfunction:
```
Comment (dbarnett) [inline autoload/codefmt.vim]: You've got a stowaway colon here.

File: autoload/codefmt.vim [hunk 5]
```diff
@@ -163,7 +191,7 @@ function! codefmt#GetClangFormatFormatter() abort
   function l:formatter.AppliesToBuffer() abort
     return &filetype is# 'c' || &filetype is# 'cpp' ||
          \ &filetype is# 'proto' || &filetype is# 'javascript'
```
Comment (dbarnett) [inline autoload/codefmt.vim]: Nit: indent +4 spaces. This seems to be 5?

**Comment by dbarnett** [comment]
LGTM.

**Pull request merged**